src/arrays/programs_involving_multidimensional_data/medium/q4/MainTest.java

```diff
@@ -0,0 +1,129 @@
+package arrays.programs_involving_multidimensional_data.medium.q4;
+
+import global.BaseRandomTest;
+import global.MethodTest;
+import global.tools.CustomAssertions;
+import global.tools.TestOption;
+import global.variables.Clause;
+import global.variables.clauses.IntegerLiteral;
+import global.variables.clauses.NewLine;
+import global.variables.clauses.RandomInteger;
+import global.variables.clauses.StringLiteral;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+import java.util.stream.Stream;
+
+import static global.utils.ArrayUtil.generateRandomArray;
+
+public class MainTest extends BaseRandomTest {
+    // Parsons with Distractors
+
+    public static final int n = 5;
+    public static final int up = 10;
+    public static final int down = 1;
+
+    public Clause[] testSentence() {
+        Clause[] c = new Clause[2 * n * n + n + 4];
+        int t = 0;
+        for (int i = 0; i < n; i++) {
+            for (int j = 0; j < n; j++) {
+                c[t++] = new RandomInteger(down, up + 1, i + " " + j);
+                c[t++] = new StringLiteral(" ");
+            }
+            c[t++] = new NewLine();
+        }
+        c[t++] = new IntegerLiteral("Even");
+        c[t++] = new NewLine();
+        c[t++] = new IntegerLiteral("Odd");
+        c[t] = new NewLine();
+        return c;
+    }
+
+    public void runMain() {
+        OddorEven.main(new String[0]);
+    }
+
+    @Test
+    public void printsCorrectOutput() {
+        int ev = 0, od = 0;
+        for (int i = 0; i < n; i++) {
+            for (int j = 0; j < n; j++) {
+                int out = Integer.parseInt(getItemByName(i + " " + j));
+                if (out % 2 == 0)
+                    ev++;
+                else
+                    od++;
+            }
+        }
+        CustomAssertions._assertArrayEquals(new int[]{ev, od}, new int[]{Integer.parseInt(getItemByName("Even")), Integer.parseInt(getItemByName("Odd"))},
+                "Your program does not correctly count the number of even and odd element in the 2D array.");
+    }
+
+    static Stream<Arguments> inputProvider() {
+        int[][] t1 = {
+                {1, 2, 3, 4, 5},
+                {2, 3, 4, 5, 1},
+                {3, 4, 5, 1, 2},
+                {4, 5, 1, 2, 3},
+                {5, 1, 2, 3, 4}
+        };
+        int[][] t2 = {
+                {1, 0},
+                {0, 1}
+        };
+        int[][] t3 = {
+                {0, 0},
+                {0, 0}
+        };
+        int[][] t4 = {
+                {2}
+        };
+        int[][] t5 = {
+                {3}
+        };
+        int[][] t6 = new int[50][50];
+        for(int i = 0; i < 50; i ++)
```

Comment:
Can you abstract this generation and counting to a helper method? It does not really belong in this method.